src/ucp/core/ucp_request.inl

```diff
@@ -394,12 +394,19 @@ ucp_request_send_state_reset(ucp_request_t *req,
     case UCP_REQUEST_SEND_PROTO_RNDV_GET:
     case UCP_REQUEST_SEND_PROTO_RNDV_PUT:
     case UCP_REQUEST_SEND_PROTO_ZCOPY_AM:
-        req->send.state.uct_comp.func   = comp_cb;
         req->send.state.uct_comp.count  = 0;
         req->send.state.uct_comp.status = UCS_OK;
         /* Fall through */
     case UCP_REQUEST_SEND_PROTO_BCOPY_AM:
-        req->send.state.dt.offset       = 0;
+        /* Always set completion function to make sure this value is initialized
+         * when doing send fast-forwarding in ucp_request_send_state_ff() */
+        req->send.state.uct_comp.func = comp_cb;
+        req->send.state.dt.offset     = 0;
+
+        /* Completion function for AM Bcopy protocol must be always NULL */
+        ucs_assert(((req->send.state.uct_comp.func == NULL) &&
+                    (proto == UCP_REQUEST_SEND_PROTO_BCOPY_AM)) ||
+                   (proto != UCP_REQUEST_SEND_PROTO_BCOPY_AM));
```

Comment:
maybe `((comp_cb == NULL) || (proto != UCP_REQUEST_SEND_PROTO_BCOPY_AM))` ?

Reply:
there are other protocols which set NULL (e.g. fragmented PUT Zcopy)

Reply:
this will not break them. I just propose the same but shorter

Reply:
yes, it should work